Crash on empty metric name

Break-out from https://github.com/prometheus/statsd_exporter/issues/190#issuecomment-474331791

https://github.com/prometheus/statsd_exporter/blob/1036439f69a70064532a1273047cf7f10712daaa/exporter.go#L283 does not check that the metric name is not empty.

With the following mapping:

```
mappings:
- match: .*_bar
  match_type: regex
  name: "${1}"
```

and any message that matches this mapping:

```
echo "foo_bar:1|c" | nc localhost 9125
```

the exporter crashes:

```
INFO[0000] Starting StatsD -> Prometheus Exporter (version=, branch=, revision=)  source="main.go:158"
INFO[0000] Build context (go=go1.12, user=, date=)       source="main.go:159"
INFO[0000] Accepting StatsD Traffic: UDP :9125, TCP :9125  source="main.go:160"
INFO[0000] Accepting Prometheus Requests on :9102        source="main.go:161"
panic: runtime error: index out of range

goroutine 1 [running]:
main.escapeMetricName(0x0, 0x0, 0x7, 0x1558faf)
	/Users/mr/src/github.com/prometheus/statsd_exporter/exporter.go:283 +0xcc
main.(*Exporter).handleEvent(0xc0000e3f28, 0x160d260, 0xc000129f00)
	/Users/mr/src/github.com/prometheus/statsd_exporter/exporter.go:336 +0x1b7
main.(*Exporter).Listen(0xc0000e3f28, 0xc0001921e0)
	/Users/mr/src/github.com/prometheus/statsd_exporter/exporter.go:308 +0x16e
main.main()
	/Users/mr/src/github.com/prometheus/statsd_exporter/main.go:213 +0x1425
```

We should reject samples that result in an empty metric name and log an error instead of crashing.

Additionally, we may want to reject mappings that reference non-existent capture groups but I'm not sure if that is feasible ahead of time.

--- a/exporter.go
+++ b/exporter.go
@@ -334,7 +334,8 @@ func (b *Exporter) handleEvent(event Event) {
 	prometheusLabels := event.Labels()
 	if present {
 		if mapping.Name == "" {
-			log.Debugf("The mapping for match \"%v\" generates an empty metric name.", mapping.Match)
+			log.Debugf("The mapping of '%s' for match '%s' generates an empty metric name", event.MetricName(), mapping.Match)
+			errorEventStats.WithLabelValues("empty_metric_name").Inc()
 			return
 		}
 		metricName = escapeMetricName(mapping.Name)
@@ -352,7 +353,7 @@ func (b *Exporter) handleEvent(event Event) {
 		// will cause the exporter to panic. Instead we will warn and continue to the next event.
 		if event.Value() < 0.0 {
 			log.Debugf("Counter %q is: '%f' (counter must be non-negative value)", metricName, event.Value())
-			eventStats.WithLabelValues("illegal_negative_counter").Inc()
+			errorEventStats.WithLabelValues("illegal_negative_counter").Inc()
 			return
 		}
 

--- a/exporter.go
+++ b/exporter.go
@@ -280,7 +280,7 @@ type Exporter struct {
 
 func escapeMetricName(metricName string) string {
 	// If a metric starts with a digit, prepend an underscore.
-	if metricName[0] >= '0' && metricName[0] <= '9' {
+	if len(metricName) > 0 && metricName[0] >= '0' && metricName[0] <= '9' {
 		metricName = "_" + metricName
 	}
 
@@ -333,6 +333,10 @@ func (b *Exporter) handleEvent(event Event) {
 	metricName := ""
 	prometheusLabels := event.Labels()
 	if present {
+		if mapping.Name == "" {
+			log.Debugf("The mapping for match \"%v\" generates an empty metric name.", mapping.Match)
+			return
+		}
 		metricName = escapeMetricName(mapping.Name)
 		for label, value := range labels {
 			prometheusLabels[label] = value

--- a/exporter_test.go
+++ b/exporter_test.go
@@ -60,7 +60,7 @@ func TestNegativeCounter(t *testing.T) {
 
 	updated := getTelemetryCounterValue(errorCounter)
 	if updated-prev != 1 {
-		t.Fatal("Empty metric name error event not counted")
+		t.Fatal("Illegal negative counter error not counted")
 	}
 }
 

--- a/exporter_test.go
+++ b/exporter_test.go
@@ -56,6 +56,38 @@ func TestNegativeCounter(t *testing.T) {
 	ex.Listen(events)
 }
 
+// TestEmptyStringMetric validates when a metric name ends up
+// being the empty string after applying the match replacements
+// tha we don't panic the Exporter Listener.
+func TestEmptyStringMetric(t *testing.T) {
+	events := make(chan Events)
+	go func() {
+		c := Events{
+			&CounterEvent{
+				metricName: "foo_bar",
+				value:      1,
+			},
+		}
+		events <- c
+		close(events)
+	}()
+
+	config := `
+mappings:
+- match: .*_bar
+  match_type: regex
+  name: "${1}"
+`
+	testMapper := &mapper.MetricMapper{}
+	err := testMapper.InitFromYAMLString(config)
+	if err != nil {
+		t.Fatalf("Config load error: %s %s", config, err)
+	}
+
+	ex := NewExporter(testMapper)
+	ex.Listen(events)
+}
+
 // TestInvalidUtf8InDatadogTagValue validates robustness of exporter listener
 // against datadog tags with invalid tag values.
 // It sends the same tags first with a valid value, then with an invalid one.
@@ -167,6 +199,7 @@ func TestEscapeMetricName(t *testing.T) {
 		"with😱emoji":              "with_emoji",
 		"with.*.multiple":         "with___multiple",
 		"test.web-server.foo.bar": "test_web_server_foo_bar",
+		"": "",
 	}
 
 	for in, want := range scenarios {

--- a/exporter_test.go
+++ b/exporter_test.go
@@ -199,7 +199,7 @@ func TestEscapeMetricName(t *testing.T) {
 		"with😱emoji":              "with_emoji",
 		"with.*.multiple":         "with___multiple",
 		"test.web-server.foo.bar": "test_web_server_foo_bar",
-		"": "",
+		"":                        "",
 	}
 
 	for in, want := range scenarios {

--- a/exporter_test.go
+++ b/exporter_test.go
@@ -52,8 +52,16 @@ func TestNegativeCounter(t *testing.T) {
 		close(events)
 	}()
 
+	errorCounter := errorEventStats.WithLabelValues("illegal_negative_counter")
+	prev := getTelemetryCounterValue(errorCounter)
+
 	ex := NewExporter(&mapper.MetricMapper{})
 	ex.Listen(events)
+
+	updated := getTelemetryCounterValue(errorCounter)
+	if updated-prev != 1 {
+		t.Fatal("Empty metric name error event not counted")
+	}
 }
 
 // TestEmptyStringMetric validates when a metric name ends up
@@ -84,8 +92,16 @@ mappings:
 		t.Fatalf("Config load error: %s %s", config, err)
 	}
 
+	errorCounter := errorEventStats.WithLabelValues("empty_metric_name")
+	prev := getTelemetryCounterValue(errorCounter)
+
 	ex := NewExporter(testMapper)
 	ex.Listen(events)
+
+	updated := getTelemetryCounterValue(errorCounter)
+	if updated-prev != 1 {
+		t.Fatal("Empty metric name error event not counted")
+	}
 }
 
 // TestInvalidUtf8InDatadogTagValue validates robustness of exporter listener
@@ -387,4 +403,13 @@ func labelPairsAsLabels(pairs []*dto.LabelPair) (labels prometheus.Labels) {
 		labels[*pair.Name] = value
 	}
 	return
+}
+
+func getTelemetryCounterValue(counter prometheus.Counter) float64 {
+	var metric dto.Metric
+	err := counter.Write(&metric)
+	if err != nil {
+		return 0.0
+	}
+	return metric.Counter.GetValue()
 }

--- a/telemetry.go
+++ b/telemetry.go
@@ -102,6 +102,13 @@ var (
 		},
 		[]string{"type"},
 	)
+	errorEventStats = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "statsd_exporter_events_error_total",
+			Help: "The total number of StatsD events discarded due to errors.",
+		},
+		[]string{"reason"},
+	)
 )
 
 func init() {
@@ -119,4 +126,5 @@ func init() {
 	prometheus.MustRegister(configLoads)
 	prometheus.MustRegister(mappingsCount)
 	prometheus.MustRegister(conflictingEventStats)
+	prometheus.MustRegister(errorEventStats)
 }

